todo_app/app.py

```diff
@@ -3,39 +3,135 @@
 
+    def render_usersadmin_response():
+        useradmin_view_model = UserAdminViewModel(users_db.get_list_of_users(), current_user)
+        return render_template('useradmin.html', view_model= useradmin_view_model)
+
+    @app.route('/login/callback', methods=['GET', 'POST'])
+    def _callback():
+        code = request.values['code']
+        token_url, headers, body = client.prepare_token_request(token_url = 'https://github.com/login/oauth/access_token', code = code )
+        token_response = requests.post(token_url, headers=headers, data=body, auth= (app.config.get('GITHUB_CLIENT_ID'), app.config.get('GITHUB_CLIENT_SECRET')),)
+        client.parse_request_body_response(token_response.text)['access_token']
+        
+        userinfo_endpoint = 'https://api.github.com/user'
+        uri, headers, body = client.add_token(userinfo_endpoint)
+        userinfo_response = requests.get(uri, headers=headers, data=body)
+        user_id = userinfo_response.json()['login']
+
+        if get_user(user_id) is None:
+            add_user(userinfo_response.json())
+        
+        user = get_user(user_id)
+        login_user(user)
+        
+        return redirect('/', code=302)
+
+    @app.route('/useradmin', methods=['GET'])
+    @login_required
+    def _useradmin():
+        if not is_permitted('admin'):
+            return redirect('/', code=302)
+        
+        return render_usersadmin_response()
+
+    @app.route('/usersetrole/<user_id>/<role>/<add_remove>', methods=['POST'])
+    @login_required
+    def _usersetrole(user_id,role,add_remove):
+        if not is_permitted('admin'):
+            return redirect('/', code=302)
+        
+        user = users_db.get_user(user_id)
+        roles = user.roles
+        if role in roles and add_remove == 'remove':
+            roles.remove(role)
+        elif role not in roles and add_remove == 'add':
+            roles.append(role)
+        else:
+            pass
+
+        users_db.modify_roles(user_id,roles)
+        return render_usersadmin_response()
+
     @app.route('/', methods=['GET'])
+    @login_required
     def _index():
         return render_index_response()    
 
     @app.route('/add', methods=['POST'])
+    @login_required
     def _add():
-        todoapi.add_item(request.form)
+        if is_permitted('writer'):
+            todoapi.add_item(request.form)
         return render_index_response()
 
     @app.route('/setstatus/<id>/<status>', methods=['POST'])
+    @login_required
     def _setstatus(id,status):
-        todoapi.modify_item(id, {'status': status})
+        if is_permitted('writer'):
+            todoapi.modify_item(id, {'status': status})
         return render_index_response()
 
     @app.route('/delete/<id>', methods=['POST'])
+    @login_required
     def _delete(id):
-        todoapi.delete_item(id)
+        if is_permitted('writer'):
+            todoapi.delete_item(id)
         return render_index_response()
 
     @app.template_filter('strftime')
```

Comment:
This function errors on invalid dates, but `/add` doesn't enforce that dates are present/valid so an empty due date can break the app. Probably best to wrap this in a try catch.

Nice use of template filters though

Reply:
Fixed!